gdpr/consent_parser.go

```diff
@@ -0,0 +1,74 @@
+package gdpr
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/prebid/go-gdpr/api"
+	"github.com/prebid/go-gdpr/vendorconsent"
+	tcf2 "github.com/prebid/go-gdpr/vendorconsent/tcf2"
+)
+
+// parsedConsent represents a parsed consent string containing notable version information and a convenient
+// metadata object that allows easy examination of encoded purpose and vendor information
+type parsedConsent struct {
+	encodingVersion uint8
+	listVersion     uint16
+	specVersion     uint16
+	consentMeta     tcf2.ConsentMetadata
+}
+
+// parseConsent parses and validates the specified consent string returning an instance of parsedConsent
+func parseConsent(consent string) (*parsedConsent, error) {
+	pc, err := vendorconsent.ParseString(consent)
+	if err != nil {
+		return nil, &ErrorMalformedConsent{
+			Consent: consent,
+			Cause:   err,
+		}
+	}
+	if err = validateVersions(pc); err != nil {
+		return nil, &ErrorMalformedConsent{
+			Consent: consent,
+			Cause:   err,
+		}
+	}
+	cm, ok := pc.(tcf2.ConsentMetadata)
+	if !ok {
+		err = errors.New("Unable to access TCF2 parsed consent")
+		return nil, err
+	}
+	return &parsedConsent{
+		encodingVersion: pc.Version(),
+		listVersion:     pc.VendorListVersion(),
+		specVersion:     getSpecVersion(pc.TCFPolicyVersion()),
+		consentMeta:     cm,
+	}, nil
+}
+
+// validateVersions ensures that certain version fields in the consent string contain valid values.
+// An error is returned if at least one of them is invalid
+func validateVersions(pc api.VendorConsents) (err error) {
+	version := pc.Version()
+	if version != 2 {
+		return fmt.Errorf("invalid encoding format version: %d", version)
+	}
+	policyVersion := pc.TCFPolicyVersion()
+	if policyVersion > 4 {
```

Comment:
Do we want to try our best guess for higher policy versions? Assume the new spec is backward compatible with the older parsers until we get a chance to address it directly?

Reply:
I believe this question was asked offline the other day by @SyntaxNode. It's a good point and perhaps worth raising with the committee. As implemented, we are in compliance with the spec. My preference would be to not block the PR for this reason and instead open a new PR to make the change if the committee feels a requirements change is needed. Thoughts?

Reply:
This is implemented per spec, but with all of us commenting on it I support changing the check to:

```
if policyVersion >= 4 {
  return 3
}
```

.. and then we can adjust it as new information is available. We missed `policyVersion` 3 and our code "just worked" by not enforcing an upper bound.